fastcrypto-tbls/src/dkg.rs

```diff
@@ -404,28 +447,49 @@ where
             })
             .collect::<HashMap<_, _>>();
 
-        for (from_sender, shares_from_sender) in shares {
+        for m in &messages.0 {
             vss_pk.add(
                 &id_to_m1
-                    .get(&from_sender)
+                    .get(&m.message.sender)
                     .expect("shares only includes shares from valid first messages")
                     .vss_pk,
             );
-            for share in shares_from_sender {
+            for share in &m.shares {
                 final_shares
                     .get_mut(&share.index)
                     .expect("created above")
                     .value += share.value;
             }
         }
 
+        // If I didn't receive a valid share for one of the verified messages (i.e., my complaint
+        // was not processed), then I don't have a valid share for the final key.
+        let shares = if messages.0.iter().all(|m| m.complaint.is_none()) {
+            Some(final_shares.values().cloned().collect())
+        } else {
+            None
+        };
+
         Output {
             nodes: self.nodes.clone(),
             vss_pk,
-            shares: final_shares.values().cloned().collect(),
+            shares,
         }
     }
 
+    /// Execute the previous two steps together.
```

Comment:
Would you ever want to do only one of the previous two steps a time, or could the separate parts be removed from the public interface and only have this one be public?